src/sparseconverter/__init__.py

```diff
@@ -0,0 +1,953 @@
+    (SPARSE_COO, SCIPY_COO): 0.02206325,
+    (SPARSE_COO, SPARSE_GCXS): 0.03331701,
+    (SPARSE_COO, SPARSE_COO): 7.2e-07,
+    (SPARSE_COO, CUPY_SCIPY_CSC): 0.03927987,
+    (CUPY_SCIPY_CSC, SCIPY_CSC): 0.01082002,
+    (CUPY_SCIPY_CSC, SPARSE_DOK): 0.81606757,
+    (CUPY_SCIPY_CSC, CUPY_SCIPY_COO): 0.02791661,
+    (CUPY_SCIPY_CSC, SCIPY_CSR): 0.00620633,
+    (CUPY_SCIPY_CSC, CUPY): 0.07202113,
+    (CUPY_SCIPY_CSC, CUDA): 0.01021739,
+    (CUPY_SCIPY_CSC, NUMPY): 0.01122473,
+    (CUPY_SCIPY_CSC, CUPY_SCIPY_CSR): 0.0032001,
+    (CUPY_SCIPY_CSC, NUMPY_MATRIX): 0.06228253,
+    (CUPY_SCIPY_CSC, SCIPY_COO): 0.0149932,
+    (CUPY_SCIPY_CSC, SPARSE_GCXS): 0.00343712,
+    (CUPY_SCIPY_CSC, SPARSE_COO): 0.12067594,
+    (CUPY_SCIPY_CSC, CUPY_SCIPY_CSC): 1.41e-06,
+}
+
+# In order to support subclasses and not check all formats each time
+# we cache dynamically which type maps to which format code
+_type_cache: Dict[type, Optional[ArrayBackend]] = {}
+
+
+def get_backend(arr: ArrayT) -> Optional[ArrayBackend]:
+    '''
+    Return the backend identifier for the given array
+    '''
+    t = type(arr)
+    backend = _type_cache.get(t, False)
+    if backend is False:
+        backend = None
+        # Make sure to check NumPy matrix first since numpy.matrix is a subclass
+        # of numpy.ndarray
+        for b in (NUMPY_MATRIX, ) + tuple(BACKENDS):
+            # Always return NUMPY for np.ndarray
+            if b == CUDA:
+                continue
+            try:
+                cls = _classes[b]
+            except (ImportError, ModuleNotFoundError):
+                # probably no CuPy
+                continue
+            if isinstance(arr, cls):
+                backend = b  # type: ignore
+                break
+        _type_cache[t] = backend
+    elif backend is True:  # Just for MyPy, it can actually never become True
+        raise RuntimeError()
+    return backend
+
+
+def get_converter(
+        source_backend: Optional[ArrayBackend], target_backend: Optional[ArrayBackend],
+        strict: bool = False
+) -> Converter:
+    identifier = (source_backend, target_backend)
+    res = _converters.get(identifier, None)
+    if strict and res is None:
+        raise ValueError(f"Could not find converter for {identifier}.")
+    elif res is None:
+        return _identity
+    else:
+        return res
+
+
+def cheapest_pair(
+        source_backends: Iterable[ArrayBackend], target_backends: Iterable[ArrayBackend]
+) -> Tuple[ArrayBackend, ArrayBackend]:
+    keys = itertools.product(source_backends, target_backends)
+    s = sorted(keys, key=lambda x: _cost[x])
+    cheapest_key = s[0]
+    return cheapest_key
+
+
+def for_backend(arr: ArrayT, backend: Optional[ArrayBackend], strict: bool = True) -> ArrayT:
+    converter = get_converter(get_backend(arr), backend, strict)
+    # print(source_backend, backend, converter)
+    return converter(arr)
+
+
+def check_shape(arr: ArrayT, shape: Iterable[int]) -> bool:
```

Comment:
Needs a docstring - just looking at the signature, I would assume that is returns `False` in the incompatible case, but it raises a `ValueError` instead. I think it should explicitly have a return type of `None`, or consistently return `True`/`False`.

Reply:
Good point, thx!